### PR TITLE
Fixed read/write of surface bits command.

### PR DESCRIPTION
**Pull request by akallabeth**
The optional field exBitmapDataHeader of TS_ BITMAP_DATA_EX was ignored.
Read and expose the data (currently unused)

May be related to #3518

### CODE REVIEW

File: libfreerdp/core/surface.c [hunk 1]
```diff
@@ -28,39 +28,74 @@
 
 #define TAG FREERDP_TAG("core.surface")
 
-static BOOL update_recv_surfcmd_surface_bits(rdpUpdate* update, wStream* s, UINT32* length)
+static BOOL update_recv_surfcmd_bitmap_header_ex(wStream* s, TS_COMPRESSED_BITMAP_HEADER_EX* header)
+{
+	if (!s || !header)
+		return FALSE;
+
```
Comment (hardening) [inline libfreerdp/core/surface.c]: these checks aren't needed

File: libfreerdp/core/surface.c [hunk 2]
```diff
@@ -28,39 +28,74 @@
...
+	return TRUE;
+}
+
+static BOOL update_recv_surfcmd_surface_bits(rdpUpdate* update, wStream* s)
+{
+	SURFACE_BITS_COMMAND* cmd = &update->surface_bits_command;
+
+	if (Stream_GetRemainingLength(s) < 20)
+		return FALSE;
```
Comment (hardening) [inline libfreerdp/core/surface.c]: I guess here the check should be either < 16 or < 16+24

Comment (akallabeth) [reply libfreerdp/core/surface.c]: right, `8`it is...

File: libfreerdp/core/surface.c [hunk 3]
```diff
@@ -126,21 +159,64 @@ int update_recv_surfcmds(rdpUpdate* update, UINT32 size, wStream* s)
...
+	Stream_Write_UINT64(s, header->tmSeconds);
+
+	return TRUE;
+}
+
+static BOOL update_write_surfcmd_bitmap_ex(wStream* s, const TS_BITMAP_DATA_EX* bmp)
+{
+	if (!s || !bmp)
+		return FALSE;
+
```
Comment (hardening) [inline libfreerdp/core/surface.c]: unneeded checks

Comment (hardening) [reply libfreerdp/core/surface.c]: oups sorry these checks are needed

File: server/Mac/mf_peer.c [hunk 4]
```diff
@@ -113,16 +113,17 @@ static void mf_peer_rfx_update(freerdp_peer* client)
...
-	cmd->height = rect.height;
-	cmd->bitmapDataLength = Stream_GetPosition(s);
-	cmd->bitmapData = Stream_Buffer(s);
+	cmd->bmp.bpp = 32;
+	cmd->bmp.codecID = 3;
+	cmd->bmp.width = rect.width;
+	cmd->bmp.height = rect.height;
+	cmd->bmp.bitmapDataLength = Stream_GetPosition(s);
+	cmd->bmp.bitmapData = Stream_Buffer(s);
 	//send
```
Comment (hardening) [inline server/Mac/mf_peer.c]: would be nice to remove the strange comments here

File: server/Sample/sfreerdp.c [hunk 5]
```diff
@@ -203,24 +203,26 @@ static BOOL test_peer_draw_background(freerdp_peer* client)
...
 	}
 	else
 	{
 		nsc_compose_message(context->nsc_context, s,
 		                    rgb_data, rect.width, rect.height, rect.width * 3);
-		cmd->codecID = client->settings->NSCodecId;
+		cmd->bmp.codecID = client->settings->NSCodecId;
 	}
 
+	memset(cmd, 0, sizeof(SURFACE_BITS_COMMAND));
```
Comment (mfleisz) [inline server/Sample/sfreerdp.c]: Won't this memset override cmd->bmp.codecID which is set before?

File: server/Sample/sfreerdp.c [hunk 6]
```diff
@@ -324,24 +326,26 @@ static void test_peer_draw_icon(freerdp_peer* client, int x, int y)
...
 		}
 		else
 		{
 			nsc_compose_message(context->nsc_context, s,
 			                    context->bg_data, rect.width, rect.height, rect.width * 3);
-			cmd->codecID = client->settings->NSCodecId;
+			cmd->bmp.codecID = client->settings->NSCodecId;
 		}
 
+		memset(cmd, 0, sizeof(SURFACE_BITS_COMMAND));
```
Comment (mfleisz) [inline server/Sample/sfreerdp.c]: Again cmd->bmp.codecID gets overridden

Comment (akallabeth) [reply server/Sample/sfreerdp.c]: @mfleisz correct, fixed that.

**Pull request merged**